src/app.js

```diff
@@ -211,40 +213,107 @@ export default function (app) {
       var realValue = element[name]
 
       if (name === "onUpdate") {
-        defer(value, element)
+          defer(value, element)
 
       } else if (value !== oldValue || realValue !== value) {
         setElementData(element, name, value, oldValue)
       }
     }
   }
 
-  function patch(parent, element, oldNode, node) {
+  function batchRemove (parent, element, node) {
+      batch.push(parent.removeChild.bind(parent, element))
+      if (node.data && node.data.onRemove) {
+        defer(node.data.onRemove, element)
+      }
+  }
+
+  function patch (parent, element, oldNode, newNode) {
     if (oldNode == null) {
-      element = parent.appendChild(createElementFrom(node))
+      var n = createElementFrom(newNode)
+      parent.insertBefore(n, element)
+      element = n
+
+    } else if (
+      newNode.tag !== oldNode.tag ||
+      typeof newNode !== typeof oldNode ||
+      typeof newNode === "string" && newNode !== oldNode
+    ) {
+      var n = createElementFrom(newNode)
+      parent.replaceChild(n, element)
+      element = n
+
+    } else if (oldNode.tag){
+      updateElementData(element, newNode.data, oldNode.data)
+      patchChildren(element, oldNode, newNode)
+    }
 
-    } else if (node == null) {
-      batch.push(function () {
-        if (oldNode.data && oldNode.data.onRemove) {
-          oldNode.data.onRemove(element)
-        }
-        parent.removeChild(element)
-      })
-    } else if (node.tag && node.tag === oldNode.tag) {
-      updateElementData(element, node.data, oldNode.data)
+    return element
+  }
+
+  function patchChildren(element, oldNode, newNode) {
+    // We need to index all keyed nodes in oldChildren in order to
+    // distinguish new keyed nodes from nodes to reuse in newChildren
+    var oldChildren = []
+    var oldKeys = {}
+    for (var i = 0; i < oldNode.children.length; i++) {
+      oldNode.children[i].data &&
```

Comment:
@zaceno Let's if instead of && here. It adds no value since uglifyjs will do and undo with this code anyway.